File: LeetCode/0130. Surrounded Regions/solution.cpp
/**
 * 58 / 58 test cases passed.
 * Runtime: 12 ms
 * Memory Usage: 9.6 MB 
 */
class Solution {
public:
    int direct[4][2] = {{1, 0}, {-1, 0}, {0, -1}, {0, 1}};
    int m, n;
    void dfs(vector<vector<char>> &board, int x, int y) {
        if (board[x][y] != 'O') return;
        board[x][y] = 'A';
        for (auto &dir: direct) {
            int xx = x + dir[0];
            int yy = y + dir[1];
            if (xx < 0 || xx >= m || yy < 0 || yy >= n) continue;
            dfs(board, xx, yy);
        }
    }
    void solve(vector<vector<char>>& board) {
        m = board.size(), n = board[0].size();
        for (int i = 0; i < m; ++ i) {
            dfs(board, i, 0);
            dfs(board, i, n - 1);
        }
        for (int i = 0; i < n; ++ i) {
            dfs(board, 0, i);
            dfs(board, m - 1, i);
        }
        for (int i = 0; i < m; ++ i) {
            for (int j = 0; j < n; ++ j) {
                if (board[i][j] == 'A') board[i][j] = 'O';
                else if (board[i][j] == 'O') board[i][j] = 'X';
            }
        }
    }
};

File: LeetCode/0130. Surrounded Regions/solution.py
"""
58 / 58 test cases passed.
Runtime: 48 ms
Memory Usage: 19.4 MB
"""
class Solution:
    def solve(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        m, n = len(board), len(board[0])
        def dfs(x, y):
            if board[x][y] != 'O': return
            board[x][y] = 'A'
            for xx, yy in [(x + 1, y), (x - 1, y), (x, y + 1), (x, y - 1)]:
                if 0 <= xx < m and 0 <= yy < n:
                    dfs(xx, yy)
        for i in range(m):
            dfs(i, 0)
            dfs(i, n - 1)
        for i in range(n):
            dfs(0, i)
            dfs(m - 1, i)
        for i in range(m):
            for j in range(n):
                if board[i][j] == 'A': board[i][j] = 'O'
                elif board[i][j] == 'O': board[i][j] = 'X'
        